tripy/tripy/frontend/tensor.py

```diff
@@ -205,51 +191,47 @@ def eval(self) -> Array:
         assert len(data) == 1, "Expects only one output from mlir_tensorrt.compiler executor"
         data = data[0]
         # Data is present now. Assign the underlying device type.
-        self.device = data.device
+        self.device = flat_ir.outputs[0].device
 
         Storage.build_internal([], [self.trace_tensor], data)
-        self.trace_tensor.shape = data.shape
         return data
 
-    def data(self) -> Array:
-        import tripy.common.datatype
-        from tripy.frontend.trace.ops.cast import cast
-
-        arr = self.eval()
-        if self.dtype in [tripy.common.datatype.float8, tripy.common.datatype.int4]:
-            arr = cast(Tensor(arr), tripy.common.datatype.float32).eval()
-        return arr
+    def tolist(self):
+        data_memref = self.eval()
+        if self.dtype in (datatype.float8, datatype.int4):
+            from tripy.frontend.trace.ops.cast import cast
 
-    def tolist(self) -> List:
-        return self.data().data()
+            data_memref = cast(Tensor(data_memref), datatype.float32).eval()
+        return memref.tolist(data_memref)
 
     def __iter__(self):
         raise TypeError("Iterating over tensors is not supported")
 
     def __repr__(self) -> str:
         # The Evaluation required before accessing self.trace_tensor.producer attributes.
-        arr = self.data()
+        arr = self.eval()
+        arr_str = memref.pretty_print_memref(arr)
         indentation = ""
         sep = ""
         if len(arr.shape) > 1 and any(dim > 1 for dim in arr.shape):
             indentation = " " * 4
             sep = "\n"
         return (
             f"tensor({sep}"
-            f"{indent(str(arr), prefix=indentation)}, {sep}"
-            f"{indent(f'dtype={arr.dtype}, loc={arr.device}, shape={arr.shape}', prefix=indentation)}"
+            f"{indent(arr_str, prefix=indentation)}, {sep}"
+            f"{indent(f'dtype={self.dtype}, loc={self.device}, shape={arr.shape}', prefix=indentation)}"
             f")"
         )
 
-    # Since the underlying data is an Array we reuse their __dlpack__() and __dlpack_device__() methods
+    # Since the underlying data is an MemRefValue we reuse their __dlpack__() and __dlpack_device__() methods
     def __dlpack__(self, stream: Any = None):
-        return self.eval().__dlpack__(stream=stream)
+        return self.eval().__dlpack__()
 
     def __dlpack_device__(self):
         return self.eval().__dlpack_device__()
 
     def __bool__(self):
-        data = self.data().data()
+        data = memref.tolist(self.eval())
```

Comment:
Why not `self.tolist()`?

Reply:
We could skip `cast` for `int4, float8 and bfloat16`, simply comparing the memory data in list format.